certz/certz.proto

```diff
@@ -453,11 +453,51 @@ message Certificate {
   // Actual certificate.
   // The exact encoding depends upon the type of certificate.
   // for X509, this should be a PEM encoded Certificate.
-  bytes certificate = 3;
+  // Deprecated in favour of certificate_type
+  bytes certificate = 3 [deprecated=true];
 
   // Optional.
   // Not needed if the device knows the private key or does not have to know it.
-  bytes private_key = 4;
+  // Deprecated in favour of private_key_type
+  bytes private_key = 4 [deprecated=true];
+
+  enum CertSource {
+    CERT_SOURCE_UNSPECIFIED = 0;
+
+    // certificates present on the device already
+    CERT_SOURCE_OIDEVID = 1;
+    CERT_SOURCE_IDEVID = 2;
+  }
+
+  oneof certificate_type {
+    // Actual certificate
+    // The exact encoding depends upon the type of certificate.
+    // for X509, this should be a PEM encoded Certificate.
+    bytes raw_certificate = 5;
+
+    // specify an alternative source for the certificate
+    CertSource cert_source = 6;
+  }
+
+  enum KeySource {
+    // KEY_SOURCE_UNSPECIFIED can be used in the case where the
+    // certificate does not need a private key (e.g. trusted certs)
+    KEY_SOURCE_UNSPECIFIED = 0;
+
+    // key is IDevID key stored in TPM
+    KEY_SOURCE_IDEVID_TPM = 1;
+
+    // key was generated with a CSR as part of rotation
+    KEY_SOURCE_GENERATED = 2;
```

Comment:
Will the device know how to find the matching generated private key?

Reply:
In the latest patch I updated this comment to say 
```
// key was generated with a CSR as part of the current Rotate RPC
```
So that there is no ambiguity about which rotation we should be considering here. 

The device should know how to find this private key, it can just retain a reference to it over the duration of the ongoing Rotate RPC.